### DISCUSSION TITLE
wrong checksum of assets from external library

### DISCUSSION THREAD

**Discussion by njhsi** [immich-app/immich#7804] [Q&A] [answered]
### The bug

Failed to search a photo from database 'assets' table by "checksum" as `#sha1sum /path/to/myphoto`

```
 user1@host12: sha1sum /import/image/2012/05/04/myphoto.jpg  #8d011d5d7f649e7aa712b5820b5d7f1eb29e53ce
 pgsql: SELECT * FROM "assets" WHERE "checksum" =decode('8d011d5d7f649e7aa712b5820b5d7f1eb29e53ce', 'hex');
```
Would anybody please to help confirm what checksum method is used for assets from external library? I was aware that it was SHA1SUM, but checking in database seems not. or anything I did not correct in the sql query?

Thanks!

### The OS that Immich Server is running on

Ubuntu

### Version of Immich Server

v1.98.1

### Version of Immich Mobile App

v1.98.1

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
original
```

### Your .env content

```Shell
original
```

### Reproduction steps

```bash
1. user1@host12:~$ sha1sum /import/image/2012/05/04/myphoto.jpg  #8d011d5d7f649e7aa712b5820b5d7f1eb29e53ce
2.  psql: SELECT * FROM "assets" WHERE "checksum" =decode('8d011d5d7f649e7aa712b5820b5d7f1eb29e53ce', 'hex');
3.
...
```

**Comment by mmomjian** [comment]
It is `sha1sum`. Can you please post the full output of your `sha1sum` command? It looks like there is a line missing or some strange formatting because it starts with `#`

**Comment by mmomjian** [comment]
I assume you connected to the immich database? What is the output of `\d` in the postgres console?

**Comment by jrasm91** [comment, marked as answer]
We probably need to document this better, but we had a discussion when we implemented libraries and basically determined that if the filesystem is the source of truth we're not going to do any asset deduping on them. The primary purpose of the hash is for that, so determined there was little value potentially hashing TB's of data for no reason. Long story, short - external library assets have a dummy hash value since it is required, but it ends up being the hash of the filepath if I remember correctly.

**Reply by njhsi** [reply]
linux command line to get sha1sum of myphoto, sha1sum is "66d5b0bfacd20b6c22a36048ff1d9792b9398311"
```
user1@host12:~$ sha1sum /import/image/2012/10/20/12080042.jpg
66d5b0bfacd20b6c22a36048ff1d9792b9398311   /import/image/2012/10/20/12080042.jpg
```

psql client cli to query db, checksum is "\x22af91765807e3f7db624e89059bf060b3bc2d0e"
```
immich=# \c immich
You are now connected to database "immich" as user "postgres".
immich=# \d
                      List of relations
 Schema  |             Name             |   Type   |  Owner   
---------+------------------------------+----------+----------
 public  | activity                     | table    | postgres
 public  | albums                       | table    | postgres
 public  | albums_assets_assets         | table    | postgres
 public  | albums_shared_users_users    | table    | postgres
 public  | api_keys                     | table    | postgres
 public  | asset_faces                  | table    | postgres
 public  | asset_job_status             | table    | postgres
 public  | asset_stack                  | table    | postgres
 public  | assets                       | table    | postgres
 public  | audit                        | table    | postgres
 public  | audit_id_seq                 | sequence | postgres
 public  | exif                         | table    | postgres
 public  | geodata_places               | table    | postgres
 public  | libraries                    | table    | postgres
 public  | migrations                   | table    | postgres
 public  | migrations_id_seq            | sequence | postgres
 public  | move_history                 | table    | postgres
 public  | partners                     | table    | postgres
 public  | person                       | table    | postgres
 public  | shared_link__asset           | table    | postgres
 public  | shared_links                 | table    | postgres
 public  | smart_info                   | table    | postgres
 public  | smart_search                 | table    | postgres
 public  | socket_io_attachments        | table    | postgres
 public  | socket_io_attachments_id_seq | sequence | postgres
 public  | system_config                | table    | postgres
 public  | system_metadata              | table    | postgres
 public  | tag_asset                    | table    | postgres
 public  | tags                         | table    | postgres
 public  | typeorm_metadata             | table    | postgres
 public  | user_token                   | table    | postgres
 public  | users                        | table    | postgres
 vectors | pg_vector_index_stat         | view     | postgres
(33 rows)

immich=# SELECT "originalPath",id,type,checksum FROM "assets" WHERE "originalFileName" LIKE '%12080042%';
             originalPath              |                  id                  | type  |                  checksum                  
---------------------------------------+--------------------------------------+-------+--------------------------------------------
 /import/image/2012/10/20/12080042.jpg | 99bce7ae-b8b7-4f53-a05a-1b6d40c1aa2c | IMAGE | \x22af91765807e3f7db624e89059bf060b3bc2d0e
(2 rows)
```

**Reply by linuxrrze** [reply]
> We probably need to document this better, but we had a discussion when we implemented libraries and basically determined that if the filesystem is the source of truth we're not going to do any asset deduping on them. The primary purpose of the hash is for that, so determined there was little value potentially hashing TB's of data for no reason. Long story, short - external library assets have a dummy hash value since it is required, but it ends up being the hash of the filepath if I remember correctly.

As I just encountered exactly that situation and wanted to verify checksum anyway, here's how to calculate the checksum:

The string used for hashing is 'path:' + originalPath

s. private async processEntity(...) in https://github.com/immich-app/immich/blob/main/server/src/services/library.service.ts:

`checksum: this.cryptoRepository.hashSha1(`path:${assetPath}`)`

**Comment by njhsi** [comment]
@jrasm91 thanks for the background info! it makes sense to me as I did not find any codes about file content checksum. 

And, it should've been more helpful if this is documented, or, better use 'filepathChecksum' rather than 'checksum' in assets table?

And, imho, it would be better to have a file checksum(xxhsum?) for every kind of source, so that we could locate or repair them globally.

**Reply by jrasm91** [reply]
Yes there is value in this aspect specifically, so it might be worth revisiting in the future.

**Reply by jrasm91** [reply]
This only applies to external libraries though. Normal, backed-up assets are stored with the sha1 hash as you would expect and your query probably would have worked correctly in that case.
